rpc/flipt/request.go

```diff
@@ -0,0 +1,192 @@
+package flipt
+
+type Requester interface {
+	Request() Request
+}
+
+// Subject represents what resource is being acted on.
+type Subject string
+
+// Action represents the action being taken on the resource.
+type Action string
+
+const (
+	SubjectConstraint   Subject = "constraint"
+	SubjectDistribution Subject = "distribution"
+	SubjectFlag         Subject = "flag"
+	SubjectNamespace    Subject = "namespace"
+	SubjectRollout      Subject = "rollout"
+	SubjectRule         Subject = "rule"
+	SubjectSegment      Subject = "segment"
+	SubjectToken        Subject = "token"
+	SubjectVariant      Subject = "variant"
+
+	ActionCreate Action = "create"
+	ActionDelete Action = "delete"
+	ActionUpdate Action = "update"
+	ActionRead   Action = "read"
+)
+
+type Request struct {
+	Subject Subject `json:"subject"`
+	Action  Action  `json:"action"`
+}
+
+func NewRequest(t Subject, a Action) Request {
+	return Request{
+		Subject: t,
+		Action:  a,
+	}
+}
+
+// Namespaces
+func (req *GetNamespaceRequest) Request() Request {
```

Comment:
I dont really love this way of adding metadata to each request type as it's easy to miss some which means authz would fail for that request. However I guess this could be mitigated with adding ITs/proper testing?

We do this similar pattern for [validation](https://github.com/flipt-io/flipt/blob/main/rpc/flipt/validation.go)

I do like it better however than depending on the GRPC `info.FullMethod` string in the GRPC interceptors and doing string comparisons, which seems to be the common approach:
examples:
- https://dev.to/techschoolguru/use-grpc-interceptor-for-authorization-with-jwt-1c5h
- https://github.com/libopenstorage/openstorage/blob/3d7c200148a18d9586811f0250b1b90f7466e69b/api/server/sdk/server_interceptors.go#L117

any ideas for a better approach that is enforced at compile time ideally?

Reply:
We had exactly the same issue in a previous role. This is not something that is perfectly achievable, i.e. tying it into a compile time contract. You ultimately end up having to remember to add it somewhere, no matter how you slice it or dice it. There will always be a way to skip it auth* in some way or another. Testing is **the only way** to ensure that expectations are met.

I played this dance with Go and gRPC a lot in my last role. The reality is, without a more generic resource model (e.g. one like k8s where resources have a consistent structure (group/version/kind and metadata) and a consistent set of verbage over those (get, list, create, update, delete, watch etc), we're likely at the mercy of doing the long and hard job of hand annotating everything.

The only other way I can think to make this something which is self extending, would be to hook into gRPC generation time, like we have with the Go client generator.

Again it wouldn't be perfect, but we can at this layer e.g. do things like say:
- Everything beneath `/api/v1/namespaces/` must be scoped to the name namespace
- Everything beneath `/api/v1/.../flags` must be scoped to the identified flag resource
- `GET`, `POST`, `PUT`, `DELETE` and so on require the equivalent scopes

Even though these map to the gateway generated code, we could take and infer the same meaning for the gRPC code and generate our own decorators, which enforce the implicit contract from the `flags.yml`.

Take the following:
https://github.com/flipt-io/flipt/blob/35243d9cccc3dca1d0e5266899f76eb7bc801661/rpc/flipt/flipt.yaml#L96-L97

We could infer an implementation of GetRollout that ensures the authenticated context can `get` flag X in namespace Y. We could do this in a generated Go decorator that calls down to another implementation of the Server.

```go
// generated decorator
func (a *AuthorizingServer) GetRollout(ctx context.Context, req *GetRolloutRequest) (*Rollout, error) {
    authz, err := GetAuthorizationFromContext(ctx)
    if err != nil {
        return nil, err
    }
    
    if !authz.Has(Scope{ResourceType: types.Flag, Resource: req.FlagKey, Namespace: req.NamespaceKey}) {
        return nil, fmt.Errorf("GetRollout(%s, %s): %w", req.NamespaceKey, req.FlagKey ErrUnauthorized)
    }
    return a.GetRollout(ctx, req)
}
```

We might find on the way that we need more metadata in that yaml file, but might get you the self managing / compile time guarantees desired. Based on the same canonical RPC definition.

Again, taking another example:
https://github.com/flipt-io/flipt/blob/35243d9cccc3dca1d0e5266899f76eb7bc801661/rpc/flipt/flipt.yaml#L72-L78

Here we might infer that `POST /api/v1/namespace/foo/flags` requires create in foo namespace.
And also that `POST /api/v1/flags` implies namespace == default.